BlueShift-iOS-SDK/BlueShift.m

```diff
@@ -779,29 +757,145 @@ - (void)displayInAppNotification {
+    void (^ fetchInAppBlock)(void) = ^ {
+        if (_inAppNotificationMananger && [[UIApplication sharedApplication] applicationState] == UIApplicationStateActive) {
+            [_inAppNotificationMananger fetchInAppNotificationsFromDataStore: BlueShiftInAppTriggerNowAndUpComing];
+        }
+    };
+    
+    if ([NSThread isMainThread] == YES) {
+        fetchInAppBlock();
+    } else {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            fetchInAppBlock();
+        });
+    }
+}
+
+
 - (void)fetchInAppNotificationFromAPI:(void (^_Nonnull)(void))success failure:(void (^)(NSError*))failure {
     if (_config.enableInAppNotification == YES && _inAppNotificationMananger) {
-        [_inAppNotificationMananger fetchLastInAppMessageIDFromDB:^(BOOL status, NSString *notificationID, NSString *lastTimestamp) {
-            if (status) {
-                [BlueshiftInAppNotificationRequest fetchInAppNotification: notificationID andLastTimestamp:lastTimestamp success:^(NSDictionary *dictionary){
-                    if ([dictionary objectForKey: kInAppNotificationContentPayloadKey]) {
-                        NSMutableArray *notificationArray = [dictionary objectForKey: kInAppNotificationContentPayloadKey];
-                        [_inAppNotificationMananger initializeInAppNotificationFromAPI:notificationArray handler:^(BOOL status) {
-                            success();
-                        }];
-                    }
-                } failure:^(NSError *error){
-                    failure(error);
-                }];
+        [BlueshiftInAppNotificationRequest fetchInAppNotificationWithSuccess:^(NSDictionary * apiResponse) {
+            [self handleInAppMessageForAPIResponse:apiResponse withCompletionHandler:^(BOOL status) {
+                success();
+            }];
+        } failure:^(NSError * error) {
+            failure(error);
+        }];
+    }
+}
+
+- (void)handleInAppMessageForAPIResponse:(NSDictionary *)apiResponse withCompletionHandler:(void (^)(BOOL))completionHandler {
+    if (apiResponse && [apiResponse objectForKey: kInAppNotificationContentPayloadKey]) {
+        NSMutableArray *inAppNotifications = [apiResponse objectForKey: kInAppNotificationContentPayloadKey];
+        if (inAppNotifications.count > 0) {
+            [_inAppNotificationMananger initializeInAppNotificationFromAPI:inAppNotifications handler:^(BOOL status) {
+                completionHandler(YES);
+            }];
+        } else {
+            completionHandler(YES);
+        }
+    } else {
+        completionHandler(NO);
+        [BlueshiftLog logInfo:@"The in-app API response is nil or does not have content attribute." withDetails:nil methodName:nil];
+    }
+}
+
+- (void)getInAppNotificationAPIPayloadWithCompletionHandler:(void (^)(NSDictionary * _Nullable))completionHandler {
+    [_inAppNotificationMananger fetchLastInAppMessageIDFromDB:^(BOOL status, NSString * notificationID, NSString * lastTimestamp) {
+        if (status) {
```

Comment:
quick doubt. what happens if the request for reading the last in-app fails? we will not return the dictionary?

Reply:
In case reading last in-app fails or no in-apps are there, fetchLastInAppMessageIDFromDB returns the `lastTimestamp` value as empty string.

Reply:
I will be adding few more checks for this in the next PR.